### ISSUE TITLE
False positive invalid-return-type with explicitly subscripted generic constructor

### ISSUE THREAD

**Issue by kang8** [astral-sh/ty#3265]
## Summary

`ty` incorrectly infers the type parameters of a generic class when constructed with an explicit type subscript inside a method that introduces a new `TypeVar`. The first type parameter gets collapsed into the second one.

## Minimal Reproduction

```python
from __future__ import annotations

from typing import Any, Callable, Generic, TypeVar

InputType = TypeVar("InputType")
IntermediateType = TypeVar("IntermediateType")
OutputType = TypeVar("OutputType")


class Pipeline(Generic[InputType, IntermediateType]):
    def __init__(
        self,
        handler: Callable[[InputType], IntermediateType],
        parent: Pipeline[Any, Any] | None = None,
    ) -> None:
        self.child: Pipeline[Any, Any] | None = None
        self.parent = parent

    def then(self, handler: Callable[[IntermediateType], OutputType]) -> Pipeline[IntermediateType, OutputType]:
        self.child = Pipeline[IntermediateType, OutputType](handler, parent=self)
        return self.child  # false positive error here
```

## Expected Behavior

No error. `self.child` was just assigned a `Pipeline[IntermediateType, OutputType]`, so returning it should satisfy the return type `Pipeline[IntermediateType, OutputType]`.

pyright reports **0 errors** on this code.

## Actual Behavior

```
error[invalid-return-type]: Return type does not match returned value
  --> ty_repro.py:33:74
   |
33 |     def then(self, handler: Callable[[IntermediateType], OutputType]) -> Pipeline[IntermediateType, OutputType]:
   |                                                                          -------------------------------------- Expected `Pipeline[IntermediateType@Pipeline, OutputType@then]` because of return type
34 |         self.child = Pipeline[IntermediateType, OutputType](handler, parent=self)
35 |         return self.child
   |                ^^^^^^^^^^ expected `Pipeline[IntermediateType@Pipeline, OutputType@then]`, found `Pipeline[OutputType@then, OutputType@then]`
```

ty incorrectly infers the type as `Pipeline[OutputType, OutputType]` instead of `Pipeline[IntermediateType, OutputType]` — the first type parameter `IntermediateType` is being replaced with `OutputType`.

## Environment

- ty 0.0.29
- Python 3.12
- macOS

**Comment by sharkdp** [comment]
Thank you for reporting this. This looks like a bug to me, potentially related to https://github.com/astral-sh/ty/issues/2817.

**Comment by sharifhsn** [comment]
This is a regression introduced by https://github.com/astral-sh/ruff/pull/24357.
